stl/inc/shared_mutex

```diff
@@ -227,48 +279,92 @@ public:
     shared_lock(mutex_type& _Mtx, defer_lock_t) noexcept
         : _Pmtx(_STD addressof(_Mtx)), _Owns(false) {} // construct with unlocked mutex
 
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, try_to_lock_t)
+    _NODISCARD_CTOR_LOCK
+    _Requires_lock_not_held_(_Mtx)
+    _When_(_Owns == true, _Acquires_shared_lock_(_Mtx))
+    _When_(_Owns == true, _Acquires_nonreentrant_lock_(_Mtx))
+    _When_(_Owns == true, _Post_same_lock_(_Mtx, *_Pmtx))
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, try_to_lock_t)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(_Mtx.try_lock_shared()) {} // construct with mutex and try to lock shared
 
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, adopt_lock_t) noexcept // strengthened
+    _NODISCARD_CTOR_LOCK
+    _Requires_shared_lock_held_(_Mtx)
+    _Post_satisfies_(_Owns == true)
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, adopt_lock_t) noexcept // strengthened
         : _Pmtx(_STD addressof(_Mtx)), _Owns(true) {} // construct with mutex and adopt ownership
 
     template <class _Rep, class _Period>
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, const chrono::duration<_Rep, _Period>& _Rel_time)
+    _NODISCARD_CTOR_LOCK //
+        _Requires_lock_not_held_(_Mtx)
+    _When_(_Owns == true, _Acquires_shared_lock_(_Mtx))
+    _When_(_Owns == true, _Acquires_nonreentrant_lock_(_Mtx))
+    _When_(_Owns == true, _Post_same_lock_(_Mtx, *_Pmtx))
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, const chrono::duration<_Rep, _Period>& _Rel_time)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(_Mtx.try_lock_shared_for(_Rel_time)) {
         // construct with mutex and try to lock for relative time
     }
 
     template <class _Clock, class _Duration>
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, const chrono::time_point<_Clock, _Duration>& _Abs_time)
+    _NODISCARD_CTOR_LOCK //
+        _Requires_lock_not_held_(_Mtx)
+    _When_(_Owns == true, _Acquires_shared_lock_(_Mtx))
+    _When_(_Owns == true, _Acquires_nonreentrant_lock_(_Mtx))
+    _When_(_Owns == true, _Post_same_lock_(_Mtx, *_Pmtx))
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, const chrono::time_point<_Clock, _Duration>& _Abs_time)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(_Mtx.try_lock_shared_until(_Abs_time)) {
         // construct with mutex and try to lock until absolute time
 #if _HAS_CXX20
         static_assert(chrono::is_clock_v<_Clock>, "Clock type required");
 #endif // _HAS_CXX20
     }
 
+    _When_(_Owns == true, _Releases_shared_lock_(*_Pmtx))
+    _When_(_Owns == true, _Releases_nonreentrant_lock_(*_Pmtx))
     ~shared_lock() noexcept {
         if (_Owns) {
             _Pmtx->unlock_shared();
         }
     }
 
-    _NODISCARD_CTOR_LOCK shared_lock(shared_lock&& _Other) noexcept : _Pmtx(_Other._Pmtx), _Owns(_Other._Owns) {
+    _NODISCARD_CTOR_LOCK
+    _Moves_lock_(*_Pmtx, *_Other._Pmtx)

```

Comment:
Seems like Static Code Analysis tool doesn't understand `_Moves_lock_` together with `std::move`.
https://developercommunity.visualstudio.com/t/_Moves_lock_-annotation-is-not-recognize/10322819

Coud not find a workaround.

Reply:
Thanks for reporting it - seems like a significant bug!